src/rt.rs

```diff
@@ -56,14 +137,54 @@ impl Rt {
     // 3. Other tasks on the `LocalSet` get a chance to run
     // 4. The sleep finishes
     // 5. The runtime pauses
-    pub(crate) fn tick(&self, duration: Duration) {
-        self.block_on(async {
+    pub(crate) fn tick(&mut self, duration: Duration) -> Result<bool> {
+        self.tokio.block_on(async {
             self.local
                 .run_until(async {
                     sleep(duration).await;
                 })
                 .await
-        })
+        });
+
+        // pull for software completion
+        match &self.handle {
+            Some(handle) if handle.is_finished() => {
+                // Consume handle to extract task result
+                if let Some(h) = self.handle.take() {
+                    match self.tokio.block_on(h) {
+                        // If the host was crashed the JoinError is cancelled, which
+                        // needs to be handled to not fail the simulation.
+                        Err(je) if je.is_cancelled() => {}
+                        res => res??,
+                    }
+                };
+                Ok(true)
+            }
+            Some(_) => Ok(false),
+            None => Ok(true),
+        }
+    }
+
+    pub(crate) fn crash(&mut self) {
+        if !self.is_host() {
+            panic!("can only crash host's software");
+        } else if self.handle.take().is_some() {
+            self.cancel_tasks();
+        };
+    }
+
+    pub(crate) fn bounce(&mut self) {
+        if !self.is_host() {
+            panic!("can only bounce host's software");
+        } else {
+            self.cancel_tasks();
```

Comment:
this code looks weird because cancel_tasks required mutable reference and borrow check was yelling at when I tried to use it inside of match block.

Reply:
Let's merge and I have an idea for this that can be a follow up.